kernel/devices/LApic.cpp

```diff
@@ -133,13 +180,20 @@ namespace Kernel::Devices
     void LApic::SendIpi(uint32_t destId, uint8_t vector)
     {
         uint32_t low = vector;
-        uint32_t high = destId << 24;
+        uint32_t high = (x2ModeEnabled ? destId : destId << 24);
 
         //everything else is fine as default here, all zeros.
-
-        WriteReg(LocalApicRegister::ICR1, high);
-        //writing to low dword sends IPI
-        WriteReg(LocalApicRegister::ICR0, low);
+        if(!x2ModeEnabled)
+        {
+            WriteReg(LocalApicRegister::ICR1, high);
+            //writing to low dword sends IPI
+            WriteReg(LocalApicRegister::ICR0, low);
+        } 
+        else
+        {
+            uint64_t x2IcrRegisterValue = ((uint64_t) high | (uint64_t) low);
+            WriteReg(LocalApicRegister::ICR0, x2IcrRegisterValue);
```

Comment:
Awesome! `BroadcastIpi()` (a few lines below this function) also uses this logic, can you update it there too? 
Also x2IcrRegisterValue should probably be `((uint64_t)high << 32 | (uint32_t)low)`, otherwise this will write the destination in the low dword.